src/shared/openssl-util.c

```diff
@@ -109,6 +110,80 @@ int rsa_pkey_to_suitable_key_size(
         return 0;
 }
 
+static int pkey_generate_volume_key_rsa(
+                EVP_PKEY *pkey,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                void **ret_savedata,
+                size_t *ret_savedata_size) {
+
+        assert(pkey);
+        assert(ret_decrypted_key);
+        assert(ret_savedata);
+        assert(ret_savedata_size);
+
+        int r = rsa_pkey_to_suitable_key_size(pkey, ret_decrypted_key_size);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine RSA public key size.");
+
+        log_debug("Generating %zu bytes random key.", *ret_decrypted_key_size);
+
+        *ret_decrypted_key = malloc(*ret_decrypted_key_size);
+        if (!*ret_decrypted_key)
+                        return log_oom();
+
+        r = crypto_random_bytes(*ret_decrypted_key, *ret_decrypted_key_size);
+        if (r < 0)
+                return log_error_errno(r, "Failed to generate random key: %m");
+
+        r = rsa_encrypt_bytes(pkey, *ret_decrypted_key, *ret_decrypted_key_size, ret_savedata, ret_savedata_size);
+        if (r < 0)
+                return log_error_errno(r, "Failed to encrypt key: %m");
+
+        return 0;
+
+}
+
+int X509_certificate_generate_volume_key(
+                X509 *cert,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                void **ret_savedata,
+                size_t *ret_savedata_size) {
+
+        assert_se(cert);
+        assert_se(ret_decrypted_key);
+        assert_se(ret_decrypted_key_size);
+        assert_se(ret_savedata);
+        assert_se(ret_savedata_size);
+
+        _cleanup_free_ void *decrypted_key = NULL;
+        size_t decrypted_key_size = 0;
+        _cleanup_free_ void *savedata = NULL;
+        size_t savedata_size = 0;
+        int r = 0;
+
+        EVP_PKEY *pkey = X509_get0_pubkey(cert);
+        if (!pkey)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to extract public key from X.509 certificate.");
+
+
+        int type = EVP_PKEY_base_id(pkey);
+        if (type == EVP_PKEY_RSA)
+                r = pkey_generate_volume_key_rsa(pkey, &decrypted_key, &decrypted_key_size, &savedata, &savedata_size);
+        else
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Unsuported public key type: %s", OBJ_nid2sn(type));
```

Comment:
we generally indent the error code paths, not the clean one, wherever it makes sense. hence, please rearrange this to:

```c
if (type != EVP_PKEY_RSA)
        return log_error_errno(…);

r = pkey_generate_volume_key_rsa(…);
if (r < 0)
        …
```

Reply:
I rewrited it with `switch`. Looks clear for me. What do you think?